libcudacxx/include/cuda/std/__utility/pod_tuple.h

```diff
@@ -0,0 +1,541 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef __CUDA_STD___UTILITY_POD_TUPLE_H
+#define __CUDA_STD___UTILITY_POD_TUPLE_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__type_traits/copy_cvref.h>
+#include <cuda/std/__type_traits/decay.h>
+#include <cuda/std/__type_traits/is_callable.h>
+#include <cuda/std/__type_traits/type_list.h>
+#include <cuda/std/__utility/integer_sequence.h>
+
+/**
+ * @file pod_tuple.h
+ * @brief Provides a lightweight implementation of a tuple-like structure that can be
+ * aggregate-initialized. It can be used to return a tuple of immovable types from a function.
+ * It is guaranteed to be a structural type up to 8 elements.
+ *
+ * This header defines the `__tupl` template and related utilities for creating and
+ * manipulating tuples with compile-time optimizations.
+ *
+ * @details
+ * The `__tupl` structure is designed to minimize template instantiations and improve
+ * compile-time performance by unrolling tuples of sizes 1-8. It also provides utilities
+ * for accessing tuple elements and applying callable objects to tuple contents.
+ *
+ * Key features:
+ * - Lightweight tuple implementation for POD types.
+ * - Compile-time optimizations for small tuples (sizes 1-8).
+ * - Support for callable application via `__apply`.
+ * - Utilities for accessing tuple elements using `__cget`.
+ * - Compatibility with CUDA device and host code.
+ */
+
+#if _CCCL_COMPILER(CLANG, <, 19)
+// See https://github.com/llvm/llvm-project/issues/88077
+#  define _CCCL_BROKEN_NO_UNIQUE_ADDRESS
+#endif
+
+#define _CCCL_API         _CCCL_HOST_DEVICE _CCCL_VISIBILITY_HIDDEN _CCCL_EXCLUDE_FROM_EXPLICIT_INSTANTIATION
+#define _CCCL_TRIVIAL_API _CCCL_API _CCCL_FORCEINLINE _CCCL_ARTIFICIAL _CCCL_NODEBUG
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+template <size_t _Idx, class _Ty>
+struct __box
+{
+  // Too many compiler bugs with [[no_unique_address]] to use it here.
+  // E.g., https://github.com/llvm/llvm-project/issues/88077
+#if !defined(_CCCL_BROKEN_NO_UNIQUE_ADDRESS)
+  _CCCL_NO_UNIQUE_ADDRESS
+#endif
+  _Ty __value_;
+};
```

Comment:
I beleive we should use `__tuple_leaf` or at least the same techniques as it uses to manually emulate EBCO

Reply:
part of the purpose of `__tuple` is to be extremely lightweight. `__tuple_leaf` is doing a lot of work to approximate `[[no_unique_address]]`. i would prefer to simply use `[[no_unique_address]]` when the compiler supports it, and otherwise accept the bloat.